# PYTHON WORKSHEET 12 (DATA CLEANING)_ANSWER KEYS

Importing Dataset

In [1]:
import pandas as pd
import numpy as np

df = pd.read_csv("data.csv")
df.head()

Suburb             Address  Rooms Type      Price Method SellerG  \
0  Abbotsford       68 Studley St    2.0    h        NaN     SS  Jellis   
1  Abbotsford        85 Turner St    2.0    h  1480000.0      S  Biggin   
2  Abbotsford     25 Bloomburg St    2.0    h  1035000.0      S  Biggin   
3  Abbotsford  18/659 Victoria St    3.0    u        NaN     VB  Rounds   
4  Abbotsford        5 Charles St    3.0    h  1465000.0     SP  Biggin   

         Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  03-09-2016       2.5    3067.0  ...       1.0  1.0     126.0           NaN   
1  03-12-2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
2  04-02-2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
3  04-02-2016       2.5    3067.0  ...       2.0  1.0       0.0           NaN   
4  04-03-2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   

   YearBuilt  CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN        Yarra  -37.8014    144.9958  Northern Metropolitan   
1        NaN        Yarra  -37.7996    144.9984  Northern Metropolitan   
2     1900.0        Yarra  -37.8079    144.9934  Northern Metropolitan   
3        NaN        Yarra  -37.8114    145.0116  Northern Metropolitan   
4     1900.0        Yarra  -37.8093    144.9944  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  
3        4019.0  
4        4019.0  

[5 rows x 21 columns]

#### 1. No. of Rows and No. of columns

In [2]:
print("No. of rows = ", df.shape[0])
print("No. of columns = ", df.shape[1])

No. of rows =  23678
No. of columns =  21


#### 2. Information about data types of each column

In [3]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23678 entries, 0 to 23677
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Suburb         23547 non-null  object 
 1   Address        23547 non-null  object 
 2   Rooms          23547 non-null  float64
 3   Type           23547 non-null  object 
 4   Price          18396 non-null  float64
 5   Method         23547 non-null  object 
 6   SellerG        23547 non-null  object 
 7   Date           23547 non-null  object 
 8   Distance       23546 non-null  float64
 9   Postcode       23546 non-null  float64
 10  Bedroom2       19066 non-null  float64
 11  Bathroom       19063 non-null  float64
 12  Car            18921 non-null  float64
 13  Landsize       17410 non-null  float64
 14  BuildingArea   10018 non-null  float64
 15  YearBuilt      11540 non-null  float64
 16  CouncilArea    15656 non-null  object 
 17  Lattitude      19243 non-null  float64
 18  Longti

#### 3. Rows having all missing values

In [4]:
# Row indexes of empty rows
print("Row indexes of empty rows : ", df[df.isnull().all(axis =1)].index.tolist())

Row indexes of empty rows :  [31, 1145, 1348, 2119, 3277, 3288, 3745, 3752, 3798, 3855, 4089, 4096, 4164, 4218, 4222, 4326, 4649, 4897, 4942, 5052, 5223, 5265, 5328, 5331, 5383, 5405, 5625, 5741, 5952, 6125, 6133, 6140, 6143, 6251, 6326, 6406, 6446, 6703, 6717, 6853, 6921, 7086, 7261, 7348, 7596, 7743, 7759, 7762, 7775, 8054, 8095, 8118, 8190, 8409, 8501, 8652, 8816, 8974, 9098, 9210, 9232, 9450, 9455, 9586, 9761, 9888, 9894, 10009, 10125, 10390, 10475, 10738, 10952, 11334, 11348, 11552, 11755, 12087, 12441, 12559, 12997, 13066, 13381, 13647, 13725, 14009, 14471, 14644, 14963, 15094, 15214, 15771, 16044, 16252, 16563, 16655, 16848, 17111, 17246, 17611, 17741, 18077, 18148, 18251, 18374, 18476, 18547, 18671, 18742, 18884, 19001, 19287, 19641, 19949, 20087, 20422, 20496, 20695, 20838, 20880, 21055, 21220, 21343, 21602, 22003, 22109, 22288, 22704, 22918, 23153, 23341]


#### 4. Number of empty rows and their percentage in whole dataset

In [5]:
print("Number of empty rows = ", df.isnull().all(axis =1).sum())
print("Percentage of empty rows in whole dataset = {}%".format(round(100*df.isnull().all(axis = 1).sum()/len(df),2)))

Number of empty rows =  131
Percentage of empty rows in whole dataset = 0.55%


#### Removing empty rows

In [6]:
df = df[~df.isnull().all(axis = 1)]
print("Number of empty rows left = ", df.isnull().all(axis =1).sum())

Number of empty rows left =  0


#### 5. Columns having missing values

In [7]:
print('Columns having missing values are as follows:')
for i in df.isnull().any().index:
    if (df.isnull().any()[i]):
        print(i)

Columns having missing values are as follows:
Price
Distance
Postcode
Bedroom2
Bathroom
Car
Landsize
BuildingArea
YearBuilt
CouncilArea
Lattitude
Longtitude
Regionname
Propertycount


#### 6. Number of missing values and their percentage in each column

In [8]:
print("Number of missing values in each column:")
print(df.isnull().sum())
print('\nPercentage of missing values in each column:')
print(round(100*df.isnull().sum()/len(df.index)),2)

Number of missing values in each column:
Suburb               0
Address              0
Rooms                0
Type                 0
Price             5151
Method               0
SellerG              0
Date                 0
Distance             1
Postcode             1
Bedroom2          4481
Bathroom          4484
Car               4626
Landsize          6137
BuildingArea     13529
YearBuilt        12007
CouncilArea       7891
Lattitude         4304
Longtitude        4304
Regionname           1
Propertycount        1
dtype: int64

Percentage of missing values in each column:
Suburb            0.0
Address           0.0
Rooms             0.0
Type              0.0
Price            22.0
Method            0.0
SellerG           0.0
Date              0.0
Distance          0.0
Postcode          0.0
Bedroom2         19.0
Bathroom         19.0
Car              20.0
Landsize         26.0
BuildingArea     57.0
YearBuilt        51.0
CouncilArea      34.0
Lattitude        18.0
Longtitude       18.0

#### 7. Column names having more than 30% missing values and removing them

From above we can see, columns **CouncilArea, YearBuilt, BuildingArea** are the columns with more than 30% missing values. So we will remove these columns.

In [9]:
# Removing columns having more than 30% missing values
df = df.drop('BuildingArea', axis=1)
df = df.drop('YearBuilt', axis=1)
df = df.drop('CouncilArea', axis=1)

round(100*(df.isnull().sum()/len(df.index)), 2)

Suburb            0.00
Address           0.00
Rooms             0.00
Type              0.00
Price            21.88
Method            0.00
SellerG           0.00
Date              0.00
Distance          0.00
Postcode          0.00
Bedroom2         19.03
Bathroom         19.04
Car              19.65
Landsize         26.06
Lattitude        18.28
Longtitude       18.28
Regionname        0.00
Propertycount     0.00
dtype: float64

#### 8. Rows with missing values

In [10]:
print("Row indexes of rows containing missing values: ", df[df.isnull().any(axis = 1)].index.tolist())

Row indexes of rows containing missing values:  [0, 3, 7, 8, 9, 12, 13, 15, 16, 19, 20, 21, 22, 26, 27, 32, 34, 35, 40, 46, 47, 48, 53, 55, 61, 63, 65, 66, 70, 75, 76, 77, 78, 79, 85, 88, 90, 92, 93, 94, 96, 98, 100, 108, 112, 116, 122, 124, 125, 126, 127, 128, 131, 132, 133, 138, 140, 144, 147, 151, 158, 161, 164, 165, 168, 170, 172, 174, 175, 176, 185, 188, 189, 190, 195, 199, 200, 201, 204, 208, 209, 212, 213, 221, 222, 223, 224, 226, 228, 241, 253, 255, 257, 267, 271, 274, 275, 284, 287, 288, 292, 294, 297, 298, 299, 301, 303, 304, 305, 306, 308, 309, 310, 312, 313, 316, 317, 318, 319, 320, 321, 324, 325, 327, 328, 329, 330, 331, 332, 333, 334, 336, 337, 342, 346, 349, 352, 353, 355, 358, 359, 363, 364, 365, 370, 371, 374, 375, 377, 379, 380, 382, 385, 386, 390, 392, 396, 397, 398, 399, 402, 405, 407, 409, 412, 413, 414, 416, 418, 420, 422, 423, 424, 425, 426, 428, 434, 435, 437, 440, 443, 444, 447, 448, 449, 451, 452, 457, 460, 461, 463, 464, 465, 470, 471, 473, 475, 476, 477, 478

#### 9. Number of missing values and their percentage in each row

In [11]:
print(pd.DataFrame.from_dict({'Rows' : df.isnull().any(axis = 1), 
                              'Number of missing values' : df.isnull().sum(axis = 1), 
                              'Percentage of missing values' : round(100*df.isnull().sum(axis = 1)/df.shape[1])})
     )

        Rows  Number of missing values  Percentage of missing values
0       True                         1                           6.0
1      False                         0                           0.0
2      False                         0                           0.0
3       True                         1                           6.0
4      False                         0                           0.0
...      ...                       ...                           ...
23673   True                         1                           6.0
23674   True                         5                          28.0
23675   True                         1                           6.0
23676  False                         0                           0.0
23677   True                         1                           6.0

[23547 rows x 3 columns]


#### 10. Number and percentage of rows having more than 5 missing values

In [12]:
# count the number of rows having > 5 missing values
print("Number of rows having more than 5 missing values : ", len(df[df.isnull().sum(axis=1) > 5].index))
print("Number of rows having more than 5 missing values : ", 100*(len(df[df.isnull().sum(axis=1) > 5].index) / len(df.index)))

Number of rows having more than 5 missing values :  4278
Number of rows having more than 5 missing values :  18.16791948018856


#### since the percentahe of such rows is less than 20%. Therefore, removing all such rows

In [13]:
# retaining the rows having <= 5 NaNs
df = df[df.isnull().sum(axis=1) <= 5]

#### 11. Description of data

In [14]:
df.describe().T

count          mean            std          min          25%  \
Rooms          19269.0  2.973117e+00       0.978969      1.00000       2.0000   
Price          15086.0  1.066695e+06  638933.149010  85000.00000  640000.0000   
Distance       19269.0  1.020436e+01       6.065786      0.00000       6.1000   
Postcode       19269.0  3.110016e+03      93.730487   3000.00000    3049.0000   
Bedroom2       19066.0  2.951956e+00       0.996032      0.00000       2.0000   
Bathroom       19063.0  1.570897e+00       0.712684      0.00000       1.0000   
Car            18921.0  1.626235e+00       0.974048      0.00000       1.0000   
Landsize       17410.0  5.517835e+02    3544.288013      0.00000     181.0000   
Lattitude      19243.0 -3.781243e+01       0.079926    -38.18418     -37.8593   
Longtitude     19243.0  1.450003e+02       0.106071    144.43162     144.9393   
Propertycount  19269.0  7.490250e+03    4319.566673    129.00000    4442.0000   

                       50%           75%           max  
Rooms               3.0000  4.000000e+00  1.200000e+01  
Price          900000.0000  1.320000e+06  9.000000e+06  
Distance            9.2000  1.300000e+01  4.810000e+01  
Postcode         3102.0000  3.150000e+03  3.977000e+03  
Bedroom2            3.0000  4.000000e+00  3.000000e+01  
Bathroom            1.0000  2.000000e+00  1.200000e+01  
Car                 2.0000  2.000000e+00  2.600000e+01  
Landsize          448.0000  6.560000e+02  4.330140e+05  
Lattitude         -37.8097 -3.775980e+01 -3.740758e+01  
Longtitude        145.0043  1.450631e+02  1.455264e+02  
Propertycount    6567.0000  1.033100e+04  2.165000e+04

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19269 entries, 0 to 23677
Data columns (total 18 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Suburb         19269 non-null  object 
 1   Address        19269 non-null  object 
 2   Rooms          19269 non-null  float64
 3   Type           19269 non-null  object 
 4   Price          15086 non-null  float64
 5   Method         19269 non-null  object 
 6   SellerG        19269 non-null  object 
 7   Date           19269 non-null  object 
 8   Distance       19269 non-null  float64
 9   Postcode       19269 non-null  float64
 10  Bedroom2       19066 non-null  float64
 11  Bathroom       19063 non-null  float64
 12  Car            18921 non-null  float64
 13  Landsize       17410 non-null  float64
 14  Lattitude      19243 non-null  float64
 15  Longtitude     19243 non-null  float64
 16  Regionname     19269 non-null  object 
 17  Propertycount  19269 non-null  float64
dtypes: flo

#### Percentage of missing values in each column:

In [16]:
print(round(100*(df.isnull().sum()/len(df.index)), 2))

Suburb            0.00
Address           0.00
Rooms             0.00
Type              0.00
Price            21.71
Method            0.00
SellerG           0.00
Date              0.00
Distance          0.00
Postcode          0.00
Bedroom2          1.05
Bathroom          1.07
Car               1.81
Landsize          9.65
Lattitude         0.13
Longtitude        0.13
Regionname        0.00
Propertycount     0.00
dtype: float64


#### 12. treating the missing values of target column Price:
Here we can see that Price column has more than 20% of missing values.
* Since this column is float type column, we can impute it by mean or median.
* But in this case if we impute such large number(21.71 %) of values in target column, our model will get biased towards mean or median value,which is not a good sign. 
* In such cases if the missing values are around 2-3% then we may impute it.
* Due to these reasons we do not have any option left except removing the rows having price column empty. So we will remove all such rows.

In [17]:
# removing NaN Price rows
df = df[~np.isnan(df['Price'])]

#### 13. Checking the percentage of missing values again in each column and treating them

In [18]:
# Percentage of missing values in each column
print(round(100*(df.isnull().sum()/len(df.index)), 2))

Suburb           0.00
Address          0.00
Rooms            0.00
Type             0.00
Price            0.00
Method           0.00
SellerG          0.00
Date             0.00
Distance         0.00
Postcode         0.00
Bedroom2         1.05
Bathroom         1.07
Car              1.76
Landsize         9.83
Lattitude        0.15
Longtitude       0.15
Regionname       0.00
Propertycount    0.00
dtype: float64


Here we can see that **Landsize** column has a significant number of missing values i.e. 9.83%. Let us see the description of this column.

In [19]:
df['Landsize'].describe()

count     13603.000000
mean        558.116371
std        3987.326586
min           0.000000
25%         176.500000
50%         440.000000
75%         651.000000
max      433014.000000
Name: Landsize, dtype: float64

Here we can see that mean and median of Lansize column are 558.17 and 440 respectively. Also, 75th percentile value is 651 but max value 433014. This represents that data has outliers. Now let us see 90th, 95th and 99th percentile values.

In [20]:
df['Landsize'].describe(percentiles = [0.9, 0.95, 0.99])

count     13603.000000
mean        558.116371
std        3987.326586
min           0.000000
50%         440.000000
90%         802.000000
95%         995.000000
99%        2954.440000
max      433014.000000
Name: Landsize, dtype: float64

In such a case ideally we must impute the values either after outlier removal. But since we do not have to remove outliers in this assignment, we are imputing the missing values by median.

In [21]:
df['Landsize'].fillna(df['Landsize'].median(), inplace = True)

# Now lets check the percentage of missing values again
print(round(100*(df.isnull().sum()/len(df.index)), 2))

Suburb           0.00
Address          0.00
Rooms            0.00
Type             0.00
Price            0.00
Method           0.00
SellerG          0.00
Date             0.00
Distance         0.00
Postcode         0.00
Bedroom2         1.05
Bathroom         1.07
Car              1.76
Landsize         0.00
Lattitude        0.15
Longtitude       0.15
Regionname       0.00
Propertycount    0.00
dtype: float64


Let us now check the distribution of **Bedroom2, Bathroom, Car, Lattitude, Longtitude** columns

In [22]:
df[['Bedroom2', 'Bathroom', 'Car', 'Lattitude', 'Longtitude']].describe()

Bedroom2      Bathroom           Car     Lattitude    Longtitude
count  14927.000000  14925.000000  14820.000000  15064.000000  15064.000000
mean       2.913043      1.538492      1.615520    -37.809849    144.996338
std        0.964641      0.689311      0.955916      0.081152      0.106375
min        0.000000      0.000000      0.000000    -38.182550    144.431810
25%        2.000000      1.000000      1.000000    -37.858100    144.931193
50%        3.000000      1.000000      2.000000    -37.803625    145.000920
75%        3.000000      2.000000      2.000000    -37.756270    145.060000
max       20.000000      8.000000     10.000000    -37.408530    145.526350

Here we can see that Bedroom2, Bathroom, Car columns consist of integer values. Lets find out by printing unique values.

In [23]:
df['Bedroom2'].unique()

array([ 2.,  3.,  4.,  1., nan,  6.,  5.,  0.,  8.,  9.,  7., 20., 10.])

In [24]:
df['Bathroom'].unique()

array([ 1.,  2., nan,  3.,  4.,  0.,  7.,  5.,  6.,  8.])

In [25]:
df['Car'].unique()

array([ 1.,  0.,  2., nan,  6.,  5.,  4.,  3.,  8.,  7.,  9., 10.])

Here we can see that all the unique values in these three columns are integers and therefore it implies that these variables are representing, Number of Bedrroms, Bathrooms or Car. So for all these columns we must impute null values using mode.

In [26]:
df['Bedroom2'] = df['Bedroom2'].fillna(df['Bedroom2'].mode().iloc[0])
df['Bathroom'] = df['Bathroom'].fillna(df['Bathroom'].mode().iloc[0])
df['Car'] = df['Car'].fillna(df['Car'].mode().iloc[0])

Now for Lattitude and Longtitude columns we may impute them by mean()

In [27]:
df['Lattitude'] = df['Lattitude'].fillna(df['Lattitude'].mean())
df['Longtitude'] = df['Longtitude'].fillna(df['Longtitude'].mean())

#### Now since all the data cleaning steps have been done. let us check the missing values in whole dataframe

In [28]:
df.isnull().sum()

Suburb           0
Address          0
Rooms            0
Type             0
Price            0
Method           0
SellerG          0
Date             0
Distance         0
Postcode         0
Bedroom2         0
Bathroom         0
Car              0
Landsize         0
Lattitude        0
Longtitude       0
Regionname       0
Propertycount    0
dtype: int64

#### 14. Number and percentage of columns removed
Initial number of columns =21

In [29]:
print("No. of columns left = ", df.shape[1])
print("No. of columns removed = ", 21 - df.shape[1])
print("Percentage of columns removed = {}%".format(round(100*(21 - df.shape[1])/21,2)))

No. of columns left =  18
No. of columns removed =  3
Percentage of columns removed = 14.29%


#### 15. Number and percentage of rows removed 

In [30]:
print("No. of Rows left = ", df.shape[0])
print("No. of Rows removed = ", 23678 - df.shape[0])
print("Percentage of Rows removed = {}%".format(round(100*(23678 - df.shape[0])/23678,2)))

No. of Rows left =  15086
No. of Rows removed =  8592
Percentage of Rows removed = 36.29%


In this exercise we got a dataset with a lot of null values and we implemented data cleaning accordingly.
* For the rows having no values at all, we have to remove them.
* For the columns having more than 30% missing values, it will have very less predictive power and we removed it for the sake of simplicity of the model training.
* For rest of the rows we saw if more than 5 columns are missing then the row has very less predictive power. Therefore we removed it.
* For the rest of the columns, we tried to take help of statistical measure and decided the treatment.

As a result we lost 36.29% of data. Although we should avoid loss of data but in case of raw datasets, these kinds of problems may arise. We cannot keep a data which is erroneous or may impart bias when training. In extreme worse cases of data, we should check our data sources also, to get a good quality of data.